include/benchmark/benchmark.h

```diff
@@ -267,12 +268,9 @@ size_t RunSpecifiedBenchmarks(BenchmarkReporter* console_reporter);
 size_t RunSpecifiedBenchmarks(BenchmarkReporter* console_reporter,
                               BenchmarkReporter* file_reporter);
 
-// If this routine is called, peak memory allocation past this point in the
-// benchmark is reported at the end of the benchmark report line. (It is
-// computed by running the benchmark once with a single iteration and a memory
-// tracer.)
-// TODO(dominic)
-// void MemoryUsage();
+// Register a MemoryManager instance that will be used to report on allocation
+// data for benchmark runs.
+void RegisterMemoryManager(MemoryManager* memory_manager);
```

Comment:
How about:
   SetMemoryManager
   GetMemoryManager

Getters are often useful to have; e.g., if somebody wants to wrap the existing memory manager.

Reply:
With the way this works, I'm not sure when they'd do that. I agree it would be useful, but it would have to happen between the registration (which should be before the call to `RunSpecifiedBenchmarks`) and the benchmark being run.

Having a const accessor might be useful if a benchmark instance wants to mutate the memory manager in some way.

If the memory manager object was part of the benchmark instance state, then it could be set on a per benchmark basis, which might be more flexible. WDYT?

Actually, thinking about it, I wonder if users could just use `UserCounters` already for this. If at the start of the instance they start their allocator measuring, and then stop it after the `KeepRunning` loop and add the allocation counters to `UserCounters`. The only issue with that might be the profiling overhead, if there is any.

Reply:
1. I  think it is fine to leave out the accesor.  But maybe rename the setter to SetMemoryManager just in case we add the getter later?

2. The main benefit of registering the memory manager outside the confines of a benchmarks is that then all benchmarks get the benefit of memory measurement automatically.   But your comment does make me think that this API is perhaps too focused one type of measurement: it is only treating one type of measurement somebody might want to add to all benchmarks.  What if somebody wants a different type of measurement; maybe they want to report contention observed in the benchmark.

So an alternative more general API might be something like:

void RegisterMetric(const string& name, Metric* metric);

class Metric {
 public:
   virtual void Start(Counter*) = 0;  // Empties counter
   virtual void Stop(Counter*) = 0;  // Fills counter
};

main (or something called at initialization time) would call RegisterMetric for all relevant metrics that should be measured across all benchmarks.  The benchmark library would call Start/Stop around benchmark execution for every registered metric.

What do you think?

Reply:
One thing I just realized is that unlike Counters, the memory measurement is collected in a separate run (because it slows down regular benchmark performance too much).  Something to keep in mind if we go with the Metric approach I proposed.  Perhaps Counter flags can be extended to indicate that the particular metric must be collected via a separate run.  Or maybe we could just do separate runs for each metric registered like this?

Reply:
Relevant: https://github.com/google/benchmark/issues/620
I wonder if timers should be generalized into metrics,
and you'd just add a second benchmark with `use timer = memory`.
I don't think counters should be taught about that.